Add testing guidelines (#5903)
prometheus-operator/prometheus-operator: TESTING.MD

Reviewer: should be named `TESTING.md` :)

Author: whoops

When contributing code to Prometheus-Operator, you'll notice that every Pull Request will run against an extensive test suite. Among an extensive list of benefits that tests brings to the Project's overall health and reliability, it can be the reviewer's and contributors's best friend during development:

* Test cases serve as documentation, providing insights into the expected behavior of the software.

> * Testing helps prevent regressions by verifying that new changes don't break existing functionality.

Reviewer: ```suggestion
* Testing can prevent regressions by verifying that new changes don't break existing functionality.
```

* Running tests locally accelerate the feedback loop, removing the dependency that contributors might have on CI when working on a Pull Request.

This document will focus on teaching you about the different test suites that we currently have and how to run different scenarios to help your development experience!

> # Test suites

Reviewer: ```suggestion
# Test categories
```

## Unit tests

Unit tests are used to test particular code snippets in isolation. They are your best ally when looking for quick feedback loops in a particular function. 

Imagine you're working on a PR that adds a new field to the Prometheus CRD and you want to test if your change is reflected to the statefulset. Instead of creating a full Kubernetes cluster, installing all the CRDs, running the Prometheus-Operator, deploying a Prometheus resource and finally check if your change made it to the live object, you could simply write or extend a unit test for the statefulset generation.

Here is an example test that checks if labels are present in the statefulset.

https://github.com/prometheus-operator/prometheus-operator/blob/aeceb0b4fadc8307a44dc55afdceca0bea50bbb0/pkg/prometheus/server/statefulset_test.go#L140-L167

Unit tests can be run with:

```
make test-unit
```

They can also be run for particular packages:

```
go test ./pkg/prometheus/server
```

Or even particular functions:

```
go test -run ^TestPodLabelsAnnotations$ ./pkg/prometheus/server
```

### Testing multiline string comparison - Golden files

> Golden files are plaintext documents designed to facilitate the validation of lengthy strings. They come in handy when, for instance, you need to test a Prometheus Config that's generated using GoLang structures. You can marshal this configuration into YAML and then compare it against a static reference to ensure a match. Golden files offer an elegant solution to this challenge, sparing you the need to hard-code the static configuration directly into your test code.

Reviewer: ```suggestion
Golden files are plaintext documents designed to facilitate the validation of lengthy strings. They come in handy when, for instance, you need to test a Prometheus configuration that's generated using Go structures. You can marshal this configuration into YAML and then compare it against a static reference to ensure a match. Golden files offer an elegant solution to this challenge, sparing you the need to hard-code the static configuration directly into your test code.
```

### Updating Golden Files

There are contributions, e.g. adding a new required field to an existing configuration, that requires us to update several golden files at once. This can easily be done with the command below:

```
make test-unit-update-golden
```

## End-to-end tests

Sometimes, running tests in isolation is not enough and we really want test the behavior of Prometheus-Operator when running in a working Kubernetes cluster. For those occasions, end-to-end tests are our choice.

To run e2e-tests, first start a Kubernetes cluter. We recommend using [MiniKube](https://minikube.sigs.k8s.io/docs/start/) or [KinD](https://kind.sigs.k8s.io/) for development purposes since they are light weight distributions of kubernetes and can run on small notebooks.

For manual e2e tests, you can use the utility script [scripts/run-external.sh](./scripts/run-external.sh), it will make sure all the requirements are in place to run Prometheus-Operator without problems, but for automated end-to-end tests, we have the command:

> 

Reviewer 2: I think it would be good to mention about the environment variables used in test suites to skip/run test suites

```bash
EXCLUDE_ALERTMANAGER_TESTS
EXCLUDE_PROMETHEUS_TESTS
EXCLUDE_PROMETHEUS_ALL_NS_TESTS
EXCLUDE_THANOSRULER_TESTS
EXCLUDE_OPERATOR_UPGRADE_TESTS
EXCLUDE_PROMETHEUS_UPGRADE_TESTS
```

Author: It is briefly mentioned in line 78. Do you feel like we could be more explicit?

Reviewer 2: I think it would be good to give example on .md file along with link. Just a thought as clicking on link can be missed and I am not sure till when the actions log will be kept

---
weight: 121
toc: true
title: Tests during development
menu:
    docs:
        parent: contributing
lead: ""
lastmod: "2023-09-10T08:48:57+00:00"
images: []
draft: false
description: Understanding and running tests when developing Prometheus-Operator

> date: "2023-09-10T00:00:00+00:00"

Author: Not sure about this part. Should we put this on the website?

Reviewer 3: I don't think so.
It will be referenced on the CONTRIBUTING and other files in same format are not part of the website

Reviewer: I don't see why it couldn't be in the website eventually but it's ok for me to not include it there in the first iteration.

Author: I've removed the headline for now

# Test suites

## Unit tests

Unit tests are used to test particular code snippets in isolation. They are your best ally when looking for quick feedback loops in a particular function. 

Imagine you're working on a PR that adds a new field to the Prometheus CRD and you want to test if your change is reflected to the statefulset. Instead of creating a full Kubernetes cluster, installing all the CRDs, running the Prometheus-Operator, deploying a Prometheus resource and finally check if your change made it to the live object, you could simply write or extend a unit test for the statefulset generation.

Here is an example test that checks if labels are present in the statefulset.

https://github.com/prometheus-operator/prometheus-operator/blob/aeceb0b4fadc8307a44dc55afdceca0bea50bbb0/pkg/prometheus/server/statefulset_test.go#L140-L167

Unit tests can be run with:

```
make test-unit
```

They can also be run for particular packages:

```
go test ./pkg/prometheus/server
```

Or even particular functions:

```
go test -run ^TestPodLabelsAnnotations$ ./pkg/prometheus/server
```

### Testing multiline string comparison - Golden files

> Golden files are plain text files used to assist assertion of long strings. In Prometheus-Operator they are often often generated by a function to be used as part of a ConfigMap or Secret.

Reviewer 3: IMO this part is a bit trick to understand, looks like we're using golden-files to store data in a config map or secret.

Maybe we can have something like that.

```suggestion
Golden files are plaintext documents designed to facilitate the validation of lengthy strings. They come in handy when, for instance, you need to test a Prometheus Config that's generated using GoLang structures. You can marshal this configuration into YAML and then compare it against a static reference to ensure a match. Golden files offer an elegant solution to this challenge, sparing you the need to hard-code the static configuration directly into your test code.
```

Author: Good suggestion, did you copy that sentence from somewhere? Looks too elegant 😛

Reviewer 3: I made a initial idea by myself and than get some help from Grammarly and ChatGPT 😛

---
weight: 121
toc: true
title: Tests during development
menu:
    docs:
        parent: contributing
lead: ""
lastmod: "2023-09-10T08:48:57+00:00"
images: []
draft: false
description: Understanding and running tests when developing Prometheus-Operator
date: "2023-09-10T00:00:00+00:00"
---

When contributing code to Prometheus-Operator, you'll notice that every Pull Request will run against an extensive test suite. Among an extensive list of benefits that tests brings to the Project's overall health and reliability, it can be the reviewer's and contributors's best friend during development:

* Test cases serve as documentation, providing insights into the expected behavior of the software.
* Testing helps prevent regressions by verifying that new changes don't break existing functionality.
* Running tests locally accelerate the feedback loop, removing the dependency that contributors might have on CI when working on a Pull Request.

This document will focus on teaching you about the different test suites that we currently have and how to run different scenarios to help you development experience!

# Test suites

## Unit tests

> Explain different unit test suites

> test short

> > test long

Reviewer: (nit) I always run all the tests and never used the short/long distinction. Not sure if it's useful to keep it.

> when to use each one

### Multiline string comparison - Golden files

Golden files are plain text files used to assist assertion of long strings. In Prometheus-Operator they are often often generated by a function to be used as part of a ConfigMap or Secret.

> In the example below, we're generating the Prometheus configuration (which can easily have 100+ lines) and comparing it against a golden file:

Reviewer: Maybe we can link to an existing code snippet?
